text/0249-v2-experiments.md

```diff
@@ -0,0 +1,137 @@
+unless they explicitly opt in by installing an experimental package or by setting a feature flag.
+
+# Motivation
+
+CDK releases contain a combination of stable and experimental features, which has proven to be a pain point for
+customers. The aws-cdk package is released frequently, at least once per week and sometimes more, and each release
+increments the minor version number (e.g 1.59.0 to 1.60.0). In the planned 2.0 release of CDK, the main focus of the
+major version upgrade is to stop packaging modules separately and to include them all in one package called aws-cdk-lib.
+This will solve a number of problems related to peer dependencies that make it harder to vend independent libraries
+based on the CDK, but it does not address the backwards compatibility problem caused by minor releases with breaking
+changes to experimental modules.
+
+The CDK uses an exception to semantic versioning expectations by labeling certain modules as experimental, to allow us
+to make breaking changes to those modules in minor version upgrades. There is precedent for this in other open source
+projects, but for CDK users, it has been a constant source of pain and frustration. Users who do not carefully read and
+understand the documentation simply install packages, copy sample code, make a few tweaks and put the code into
+production. When they later upgrade to a new version, they are surprised to find that their code no longer works. The
+perception of instability has driven some users away, or caused them to limit their usage of CDK to the fundamental L1
+constructs, which do not provide them with the benefits of higher level abstractions.
+
+This RFC proposes that we stop releasing breaking changes in the main packages that we vend that are not hidden by a
+feature flag. A user that installs aws-cdk-lib using npm or pip or any other package manager should be confident that
+there will be no intentional breaking changes in the 2.x line of releases for its lifetime.
+
+# README
+
+The following section will be removed from the CDK Readme:
+
+> Modules in the AWS Construct Library are designated Experimental while we build them; experimental modules may have
+> breaking API changes in any release. After a module is designated Stable, it adheres to semantic versioning, and only
+> major releases can have breaking changes. Each module's stability designation is available on its Overview page in the
+> AWS CDK API Reference. For more information, see Versioning in the CDK Developer Guide.
+
+The existing contributor guide does not mention experimental code, so the following will be added after the intro
+paragraph:
+
+> If you are contributing to a 2.x release, keep in mind that you may not introduce breaking changes to any public
+> facing API. Experimental modules are developed in a separate package, and experimental APIs in stable modules are
+> hidden behind a feature flag. [link]
```

Comment:
We can't really "hide" the experimental APIs, only document them and issue runtime errors if users did not opt-in.

Also, we need to decide if we want a single feature flag for all experimental APIs, or a flag per API. The former is likely the pragmatic choice. The latter will incur quite a lot of overhead I believe.

Reply:
It might not always be practical to enable them all at once. What if two experimental features conflict with each other?

Reply:
What's the dev ex for enabling these experimental APIs? How do we separate framework usage of experimental APIs from consumer usage of these APIs?

Reply:
I can't think of a good example, but for the sake of simplicity we should go with a single flag, and if any weird situations come up where that doesn't work, we can introduce new flags on an ad-hoc basis if absolutely necessary.